### PR TITLE
merge of AB test + FF + goal to develop with experiments

### PR DESCRIPTION
**Pull request by sufka861**
allready resolved merge conflicts togethere with Peer

### CODE REVIEW

File: Service/feature.logic.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+
+const featureCheckAttributes =(endUserReq, experiment) =>{
+
+
+    const {ON, OFF} = experiment.variants_ff;
+    if (Util.checkAttributes(endUserReq, experiment))
+        return {"ON": ON};
+    else
+        return {"OFF": OFF};s
```
Comment (AtomicManic) [inline Service/feature.logic.js:13]: so the logic here is if the attributes match we show on, and otherwise we show off?

File: Service/utils.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+const getClientIP =(endUserReq)=> {
+    const result = requestIp.getClientIp(endUserReq);
+    if (!result) throw new ServerUnableError("getClientIP");
+    return result;
+}
+
+const getLocation = (req)=> {
+    if (!req) throw new PropertyNotFound("getlocation");
+    const result = geoip.lookup(req.clientIp);
```
Comment (AtomicManic) [inline Service/utils.js:23]: add validation on req.clientIp

File: Service/utils.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    if (!result) throw new ServerUnableError("getClientIP");
+    return result;
+}
+
+const getBrowserDevice = (req)=>{
+    const userAgentInfo = parser(req.headers['user-agent']);
```
Comment (AtomicManic) [inline Service/utils.js:29]: validation on req.headers missing

File: repositories/experiment.repository.js [hunk 4]
```diff
@@ -1,40 +1,50 @@
...
-        return user;
+    incVariantSuccessCount(id, variant){
+        return this.update(id, {$inc : {[`variant_success_count.${variant}`]: 1}});
     }
 
-    update(id, experiment) {
-        return this.storage.update(id, experiment);
+    getVariantSuccessCount(id){
+        const experiment = this.retrieve(id);
+        if(!! experiment)
```
Comment (AtomicManic) [inline repositories/experiment.repository.js:28]: remove one !

File: repositories/experiment.repository.js [hunk 5]
```diff
@@ -1,40 +1,50 @@
...
+            return experiment.variant_success_count;
+        else
+            return null;
```
Comment (AtomicManic) [inline repositories/experiment.repository.js:31]: instead of returning null, you can do the check if(!experiment) throw new NotFoundEntity('experiment')

**Review by AtomicManic — APPROVED** [review]
Looking good. Made some comments to fix, but it's approved so after this you can merge

File: Service/feature.logic.js (hunk 1 above)
Comment (sufka861) [reply Service/feature.logic.js:13]: if the attributes match - then show the feature "ON"
else
"OFF"

**Comment by sufka861** [comment]
all coments and bugs fixed

**Pull request merged**